CONTRIBUTING.md

```diff
@@ -67,6 +67,116 @@ Some characteristics of good tests:
 * be minimal, deterministic, stable (unaffected by irrelevant changes), easy to understand and review
 * have minimal dependencies: a compiler bug test should not depend on, e.g., the Scala library
 
+There are mainly three kinds of tests: unit tests, property-based tests, and integration tests.
+
+#### JUnit
+
+For unit tests we use JUnit, which can be run from sbt shell as follows:
+
+```
+root> junit/testQuick
+```
+
+It might take a few minutes the first time you run `junit/testQuick`, but from the second time onwards
+sbt will only run the tests that is affected by the code change since the last run.
+See `test/junit/` for the examples of JUnit tests.
+
+#### ScalaCheck
+
+For testing that can benefit from having lots of randomly generated data, property-based testing should be used.
+This is run from sbt shell as follows:
+
+```
+root> scalacheck/testOnly ByOneRangeTest
+```
+
+See `test/scalacheck/range.scala`.
+
+#### Partest
```

Comment:
I think most of the time people will be working with partest, so it might be an idea to list it ahead of JUnit and ScalaCheck. For me the typical process is,
1. Copy and paste a failing case into a new (usually pos) test case.
2. Minimize.
3. Fix bug.
4. ...
5. PR!

Reply:
Partest is so powerful. But as I mentioned in there if we all kept adding it, it would make the build time longer, so I figured that the core maintainers would want to encourage more JUnit or ScalaCheck.

Reply:
I think more JUnit and ScalaCheck are encouraged, but I think partest is still by far the primary route in to fixes for external contributors: typically you start with some code which you think _should_ compile and go from there ... that's straightforward with partest, much less so otherwise.

Reply:
> I think partest is still by far the primary route in to fixes for external contributors

hopefully some better contribution guidelines can change this, so that every feature, enhancement and bugfix doesn't impose such increase in cost to testing, particularly for library changes.

for instance it would be good (eventually - I don't meant to feature creep this PR) to mention how things like `scala.tools.testing.BytecodeTesting` can be used (as Lukas used in #6343).

Reply:
JUnit and ScalaCheck for library changes seems reasonable. But for compiler changes I'd say not so much. Perhaps it would make sense to flag up that contrast? "If you're fixing a bug in the compiler use partest. If you're fixing a bug in the library use JUnit and/or ScalaCheck".

Reply:
I've added JUnit tests myself, but for me that hasn't been typical because to write a JUnit test you already have to be very familiar with the relevant internals.

That's unlikely to be the case if you're starting from "I have this Scala code which I think should compile but doesn't ...".

Reply:
This is true, but if you're already fiddling with the compiler APIs to fix one compilation issue you'll be more or less familiar with them, so unit testing may be easier and have a tighter edit compile and test than running partest, which requires publishing. I agree with @dwijnand that, in the future, it would be great if unit testing is better explained.

Reply:
There's some work being done on scala/scala-partest#75 by @retronym which should make writing partests not quite expensive on build time.

Reply:
> "If you're fixing a bug in the compiler use partest. If you're fixing a bug in the library use JUnit and/or ScalaCheck".

I like it, the definitive tone just needs to be toned down, by adding "typically" or such words

Reply:
Perhapy "... typically start with ...".